packages/flutter/lib/src/rendering/binding.dart

```diff
@@ -192,23 +196,15 @@ class FlutterBinding extends HitTestTarget {
     _renderView.compositeFrame();
   }
 
-  final List<EventListener> _eventListeners = new List<EventListener>();
-
-  /// Calls listener for every event that isn't localized to a given view coordinate
-  void addEventListener(EventListener listener) => _eventListeners.add(listener);
-
-  /// Stops calling listener for every event that isn't localized to a given view coordinate
-  bool removeEventListener(EventListener listener) => _eventListeners.remove(listener);
+  void _handlePushRoute(String route) {
+    print("_handlePushRoute $route");
+    for (BindingObserver observer in _observers)
+      observer.onPushRoute(route);
+  }
 
-  // TODO(abarth): The engine should give us the timeStamp in Durations.
-  void _handleEvent(String eventType, double timeStamp) {
-    assert(eventType == 'back');
-    InputEvent ourEvent = new InputEvent(
-      type: eventType,
-      timeStamp: new Duration(microseconds: timeStamp.round())
-    );
-    for (EventListener listener in _eventListeners)
-      listener(ourEvent);
+  void _handlePopRoute() {
+    for (BindingObserver observer in _observers)
+      observer.onPopRoute();
```

Comment:
That's scary. What if they all do something?

Maybe there should only be one observer? (Like for Navigator currently?)